staging/src/k8s.io/kubectl/pkg/kuberc/kuberc.go

```diff
@@ -0,0 +1,315 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kuberc
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/homedir"
+	"k8s.io/kubectl/pkg/cmd/util"
+	"k8s.io/kubectl/pkg/config/v1alpha1"
+	"k8s.io/kubectl/pkg/scheme"
+)
+
+const RecommendedKubeRCFileName = ".kuberc"
+
+var (
+	RecommendedConfigDir  = filepath.Join(homedir.HomeDir(), clientcmd.RecommendedHomeDir)
+	RecommendedKubeRCFile = filepath.Join(RecommendedConfigDir, RecommendedKubeRCFileName)
+)
+
+// PreferencesHandler is responsible for setting default flags
+// arguments based on user's kuberc configuration.
+type PreferencesHandler interface {
+	AddFlags(flags *pflag.FlagSet)
+	ApplyOverrides(rootCmd *cobra.Command, args []string, errOut io.Writer) error
+	ApplyAliases(rootCmd *cobra.Command, args []string, errOut io.Writer) ([]string, error)
+}
+
+// Preferences stores the kuberc file coming either from environment variable
+// or file from set in flag or the default kuberc path.
+type Preferences struct {
+	GetPreferencesFunc func(kuberc string) (*v1alpha1.Preferences, error)
+
+	aliases map[string]struct{}
+}
+
+// NewPreferences returns initialized Prefrences object.
+func NewPreferences() PreferencesHandler {
+	return &Preferences{
+		GetPreferencesFunc: DefaultGetPreferences,
+		aliases:            make(map[string]struct{}),
+	}
+}
+
+// AddFlags adds kuberc related flags into the command.
+func (p *Preferences) AddFlags(flags *pflag.FlagSet) {
+	if util.KubeRC.IsEnabled() {
+		flags.String("kuberc", "", "Path to the kuberc file to use for preferences.")
+	}
+}
+
+// ApplyOverrides finds the command and sets the defaulted flag values in kuberc.
+func (p *Preferences) ApplyOverrides(rootCmd *cobra.Command, args []string, errOut io.Writer) error {
+	if len(args) <= 1 {
+		return nil
+	}
+
+	if !util.KubeRC.IsEnabled() {
+		return nil
+	}
+
+	kubercPath := getExplicitKuberc(args)
+	kuberc, err := p.GetPreferencesFunc(kubercPath)
```

Comment:
Is there any way to avoid having to read the kuberc file multiple times (in both ApplyAliases and ApplyOverrides)?

Maybe just have a single `pref.Apply()` func that reads the kuberc file once and passes it to internal `applyAliases()` and `applyOverrides()` funcs (just an idea)?

Or have some way to reuse the contents of a previously read kuberc file.

Reply:
`pref` can easily instantiate single kuberc file read during initialization and use it in both functions.

Reply:
> Maybe just have a single pref.Apply() func that reads the kuberc file once and passes it to internal applyAliases() and applyOverrides() funcs (just an idea)?

This sounds reasonable and I changed in that way.